### PR TITLE
Add method to list liked repos

### PR DESCRIPTION
**Pull request by Wauplin**
Resolve https://github.com/huggingface/huggingface_hub/issues/1137 (cc @creatorrr @camenduru) thanks to @coyotte508 implementation ! :fire: :tada: 
(see https://github.com/huggingface/moon-landing/pull/4772 -internal link-).
Also resolve https://github.com/huggingface/huggingface_hub/issues/1255.

Add a `list_liked_repos` endpoint that returns a `UserLikes` object containing the total count and 3 lists of repo ids for models, datasets and spaces. Can list likes from any user (not necessarily yourself):
```python
>>> from huggingface_hub import HfApi
>>> likes = HfApi().list_liked_repos("julien-c")
>>> likes.user
"julien-c"
>>> likes.total
626
>>> likes.models  # or likes.datasets or likes.spaces
["osanseviero/streamlit_1.15", "Xhaheen/ChatGPT_HF", ...]
```

**Note:** the "createdAt" field is lost in the Python object. Not sure it's the most useful information in this context. If some users really want this information, they can always fetch it from the http endpoint directly.

**EDIT:**
- added `like` and `unlike` endpoints as well
- show total likes + total visible likes in UserLikes object

___

@coyotte508 on the endpoint you make a difference between `totalCount` and `visibleLikes`. In which case can we have a like that is not public ? I tried but private repos cannot be liked. If I look at @julien-c likes (https://huggingface.co/api/users/julien-c/likes), total count is 637 but only 626 repos are returned.

If some likes are privates for some reason, I am not sure it makes sense to return them in the totalCount. Otherwise if there is a good reason, I'll add an explanation to the `list_liked_repos` docstring.

```py
import requests
data =  requests.get("https://huggingface.co/api/users/julien-c/likes").json()

# 637
print(data["totalCount"])

# 626
print(len(data["visibleLikes"]))
```

### CODE REVIEW

**Comment by coyotte508** [comment]
Some repos can be liked & then become private.

Also it's only the UI that prevents liking private repos, not the backend.

As for whether it makes sense or not, on the user profile (PR on moonlanding by @krampstudio ) we'll display the total of likes, private & non-private.

File: src/huggingface_hub/__init__.py [hunk 1]
```diff
@@ -52,6 +52,9 @@
 # WARNING: any comment added in this dictionary definition will be lost when
 # re-generating the file !
 _SUBMOD_ATTRS = {
+    "_activity_api": [
+        "UserLikes",
+    ],
```
Comment (coyotte508) [inline src/huggingface_hub/__init__.py:57]: not sure about `activity_api`, cc @LysandreJik

**Comment by Wauplin** [comment]
Ah yeah, got it. I'll make the `UserLikes` more explicit then.

Out of curiosity, is there an endpoint to like a repo ? Could make sense to integrate it in `huggingface_hub` as some users seem to be interested in building list of "favorite repos" as a wishlist/remember list.

**Review by coyotte508 — COMMENTED** [review]
API call is ok!

**Comment by coyotte508** [comment]
> Out of curiosity, is there an endpoint to like a repo ? Could make sense to integrate it in `huggingface_hub` as some users seem to be interested in building list of "favorite repos" as a wishlist/remember list.

`/api/:repoType(models|spaces|datasets)/:namespace?/:repo/like` :)

**Comment by camenduru** [comment]
very good news 😭😭😭😭 thanks @Wauplin thanks @coyotte508 ❤❤❤❤

File: src/huggingface_hub/__init__.py (hunk 1 above)
Comment (Wauplin) [reply src/huggingface_hub/__init__.py:57]: yep, trying to move data structures out of hf_api (that is really getting close to 4k LOC now with a lot of stuff not directly related to the Client itself). But in this case I agree it's quite empty

**Comment by Wauplin** [comment]
Based on [@coyotte508 's comment ](https://github.com/huggingface/huggingface_hub/pull/1254#issuecomment-1343027207) I've added 2 `like` and `unlike` endpoint to `HfApi`. Also made the distinction between `total` and `total_visible`.

PR is ready for review but shouldn't not be merged before https://github.com/huggingface/moon-landing/pull/4813 (internal link) is merged => will use DELETE /like instead of POST /unlike in the future.

File: docs/source/package_reference/hf_api.mdx [hunk 2]
```diff
@@ -62,6 +62,10 @@ hf_api = HfApi(
 
 [[autodoc]] huggingface_hub.hf_api.CommitInfo
 
+### UserLikes
+
+[[autodoc]] huggingface_hub._activity_api.UserLikes
```
Comment (osanseviero) [inline docs/source/package_reference/hf_api.mdx:67]: Remember to update the toctree for it to show

File: src/huggingface_hub/hf_api.py [hunk 3]
```diff
@@ -1194,6 +1195,197 @@ def list_spaces(
...
+        #       'repo': {
+        #           'name': 'PaddlePaddle/PaddleOCR',
+        #           'type': 'space'
+        #        }
+        #   }
+        # Let's loop 3 times over the received list. Less efficient but more straightforward to read.
+        return UserLikes(
+            user=user,
+            total=data["totalCount"],
+            total_visible=len(data["visibleLikes"]),
```
Comment (osanseviero) [inline src/huggingface_hub/hf_api.py:1371]: I find this structure/API slightly uninuitive. 

`total` will give you a number, but `len(models)` will give you a different number. Maybe `total` should be the visible ones, and then we could have a `total_private`? But I'm really not sure; no strong thoughts

**Review by LysandreJik — APPROVED** [review]
This looks good! Thanks for working on it @Wauplin 

Side note, unrelated to the PR but that I'm seeing when playing with the methods: the `validate_hf_hub_args` seems to hide the true values of the expected arguments. When using it, my IDE only offers the `*args, **kwargs` instead of the `user=None, *, token=None`

File: src/huggingface_hub/hf_api.py (hunk 3 above)
Comment (julien-c) [reply src/huggingface_hub/hf_api.py:1371]: or just not expose either total? (and let users do `len(likes)`)

Comment (Wauplin) [reply src/huggingface_hub/hf_api.py:1371]: I also feel it's weird to have 2 totals but I didn't come up with a "better" way. With the current implementation `UserLikes.total` is equal to the number we can now see on everyone's profile (for example [1106 for you](https://huggingface.co/datasets/open-source-metrics/hf-doc-google-analytics)) which I think would be the biggest source of confusion if it wasn't the case.

The `total_visible` is more there to remind users that what then can see is less that the real total (1092 in your case).

File: src/huggingface_hub/_activity_api.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+# coding=utf-8
+# Copyright 2019-present, the HuggingFace Inc. team.
```
[the remark this answers is not in the document]
Comment (Wauplin) [reply src/huggingface_hub/_activity_api.py:2]: ahahah

File: src/huggingface_hub/hf_api.py (hunk 3 above)
Comment (julien-c) [reply src/huggingface_hub/hf_api.py:1371]: i think we should just align on the hub side to only display "visible likes" cc @gary149 @coyotte508 @krampstudio what do you say?

File: src/huggingface_hub/__init__.py (hunk 1 above)
Comment (Wauplin) [reply src/huggingface_hub/__init__.py:57]: > Would it make sense to move the existing dataclasses to a specific module instead? (things like model info, commit info, etc)

yep ok with this. I'll do it in a separate PR

Comment (Wauplin) [reply src/huggingface_hub/__init__.py]: For now I just removed it and put the dataclass in `hf_api.py`

File: src/huggingface_hub/hf_api.py (hunk 3 above)
Comment (julien-c) [reply src/huggingface_hub/hf_api.py:1402]: yes but again we will likely change this on the hub side anyways

File: src/huggingface_hub/__init__.py (hunk 1 above)
Comment (coyotte508) [reply src/huggingface_hub/__init__.py]: good call, that was it stays flexible for the future!

File: src/huggingface_hub/hf_api.py (hunk 3 above)
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py]: I removed the `total` and `total_visible` in https://github.com/huggingface/huggingface_hub/pull/1254/commits/bffb3a3ef4ec726501f4692ac3b672e71f2cf15d. At least we can merge this PR and potentially put back a property once this is figured out in https://github.com/huggingface/moon-landing/issues/4850 -if ever useful-

**Comment by Wauplin** [comment]
@LysandreJik @julien-c @osanseviero I addressed all the comments (removed the `_activity_api` module and removed the `total`/`total_visible` properties). Should be good to go now ?

**Review by LysandreJik — APPROVED** [review]
Thanks for adapting!

**Pull request merged**